src/pip/_internal/download.py

```diff
@@ -933,6 +933,29 @@ def unpack_http_url(
             os.unlink(from_path)
 
 
+def _copy_source_tree(source, target):
+    # type: (str, str) -> None
+    try:
+        shutil.copytree(source, target, symlinks=True)
+    except shutil.Error as e:
+        errors = e.args[0]  # type: List[Tuple[str, str, shutil.Error]]
+        # Users may have locally-created socket files in the source
+        # directory. Copying socket files is not supported, so we skip errors
+        # related to them, for convenience.
+
+        # Copy list to avoid mutation while iterating.
+        errors_copy = errors[:]
+        for i, err in reversed(list(enumerate(errors_copy))):
+            src, _dest, _error = err
+            if is_socket(src):
+                # Remove errors related to sockets to prevent distractions if
+                # we end up re-raising.
+                errors.pop(i)
```

Comment:
It seems like this more complicated error-handling logic could be eliminated if you pass as the [`copy_function` argument](https://docs.python.org/3/library/shutil.html#shutil.copytree) to `copytree()` a wrapped version of `copy2()` (the default value) that ignores errors caused by socket files. Then your error handling only needs to handle one path at a time, and you handle it before it gets added to the list, instead of the error getting added and then you needing to later remove it.

Reply:
Ah, `copytree` does not take `copy_function` in Python 2. That's unfortunate.

We could similarly use the `ignore` parameter, but I didn't want to use that and complicate efforts like #6770/.pipignore.

Reply:
If fixing this in Python 2 means a more complicated implementation, I think it’s okay / would be better to limit the fix to Python 3. This seems like a somewhat obscure bug anyways.

Reply:
I simplified this as follows:

1. Wrap `copytree()` and map socket errors to SpecialFileError ([link](https://github.com/pypa/pip/pull/6802/files#diff-6d5635c4f949690f8e6196ae6b155582R35-R61)). This is exactly the same as `shutil.copytree()`, just fixing the bug as stated in the docstring.
2. Filter out SpecialFileError errors and raise a new exception from any remaining errors ([link](https://github.com/pypa/pip/pull/6802/files#diff-34481ab2fefef95e8820a46e86004ae7R991-R996)).